xds/src/test/java/io/grpc/xds/LocalityStoreTest.java

```diff
@@ -842,6 +865,393 @@ public void reset() {
     verify(loadStatsStore, times(2)).removeLocality(locality2);
   }
 
+  /**
+   * Tests the scenario of the following sequence of events.
+   * (In the format of "event detail - expected state", with abbreviations C for CONNECTING,
+   * F for TRANSIENT_FAILURE, R for READ, and D for deactivated etc.)
+   * EDS update: P0 sz1; P1 sz2; P2 sz3; P3 sz4 - P0 C, P1 N/A, P2 N/A, P3 N/A
+   * 10 secs passes P0 still CONNECTING - P0 C, P1 C, P2 N/A, P3 N/A
+   * 5 secs passes P1 in TRANSIENT_FAILURE - P0 C, P1 F, P2 C, P3 N/A
+   * 4 secs passes P2 READY - P0 C, P1 F, P2 R, P3 N/A
+   * P0 gets READY - P0 R, P1 F&D, P2 R&D, P3 N/A
+   * P1 gets READY - P0 R, P1 R&D, P2 R&D, P3 N/A
+   * 10 min passes P0 in TRANSIENT_FAILURE - P0 F, P1 R, P2 R&D, P3 N/A
+   * 5 min passes - P0 F, P1 R, P2 N/A, P3 N/A
+   * P1 in TRANSIENT_FAILURE - P0 F, P1 F, P2 C, P3 N/A
+   * 10 secs passes - P0 F, P1 F, P2 C, P3 C
+   * P1, P3 gets READY - P0 F, P1 R, P2 C&D, P3 R&D
+   * EDS update, localities moved: P0 sz1, sz3; P1 sz4; P2 sz2 - P0 C, P1 R, P2 R&D
+   * 15 min passes - P0 C, P1 R, P2 N/A
+   * EDS update, locality removed: P0 sz1, sz3, - P0 C, P1 N/A, sz4 R&D
+   * sz3 gets READY - P0 R, P1 N/A, sz4 R&D
+   * EDS update, locality comes back and another removed: P0 sz1, P1 sz4 - P0 C, P1 R, sz3 R&D
+   */
+  @Test
+  public void multipriority() {
+    // EDS update: P0 sz1; P1 sz2; P2 sz3; P3 sz4 - P0 C, P1 N/A, P2 N/A, P3 N/A
+    LocalityInfo localityInfo1 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint11, lbEndpoint12), 1, 0);
+    LocalityInfo localityInfo2 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint21, lbEndpoint22), 2, 1);
+    LocalityInfo localityInfo3 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint31, lbEndpoint32), 3, 2);
+    LocalityInfo localityInfo4 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint41, lbEndpoint42), 3, 3);
+    final Map<XdsLocality, LocalityInfo> localityInfoMap = ImmutableMap.of(
+        locality1, localityInfo1, locality2, localityInfo2, locality3, localityInfo3, locality4,
+        localityInfo4);
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        localityStore.updateLocalityStore(localityInfoMap);
+      }
+    });
+    assertThat(loadBalancers.keySet()).containsExactly("sz1");
+    LoadBalancer lb1 = loadBalancers.get("sz1");
+    InOrder inOrder = inOrder(lb1, helper);
```

Comment:
You don't really need `inOrder`. `inOrder` verifies the sequence of interactions on one or a set of mocks, but here you are not consistently verifying the invocation sequence but keep recreating new `inOrder`s. This usage is not proper. You want to do a step-by-step check that how each child balancer helper is invoked, but you are already verifying that they are spawned one by one. So, do use `inOrder`.

Reply:
For each new `inOrder`, I'm consistently verifying the invocation order, until it's replaced. I need to verify the order of `helper.updateBalancingState` and `lb.handleResolvedAddressGroup`

Reply:
No, the order of those two methods being invoked does not matter in terms of behavior.

Reply:
The order is intended behavior, because I have 
```
          // In extreme case handleResolvedAddresses() may trigger updateBalancingState()
          // immediately, so execute handleResolvedAddresses() after all the setup in this method is
          // complete.
          helper.getSynchronizationContext().execute(new Runnable() {
            @Override
            public void run() {
              localityLbInfo.childBalancer.handleResolvedAddresses(
                  ResolvedAddresses.newBuilder().setAddresses(localityInfo.eags).build());
            }
          });
```

Reply:
That's implementation detail, i.e., __how__ you update backend addresses. The behavior you see is _backend addresses get updated_. You don't insist on the ordering of  `helper.updateBalancingState` and `lb.handleResolvedAddressGroup` as long as they are doing the desired thing.

Reply:
No, I insist `updateBalancingState` triggered by `PriorityManager` happen first, because I don't want `handleResolvedAddressGroup` happen first and trigger its own `updateBalancingState` and picker in a half-baked state.

Reply:
Here you want to verify what happens when failing over to a uninitialized locality. You want to verify the state of that locality becomes `CONNECTING`, not even bother to verify that its backend addresses are updated first. Because it's your implementation choose to not have an unused locality holding backend addresses. It's also not wrong if you each locality holds backend addresses right after locality info is updated. Verifying the strict ordering of how these methods are called is dangerous, as you are writing tests by simulating how your implementation flows. This doesn't test anything, but just runs the implementation in test.

Reply:
This is the only way I can verify the order I intended to verify. I did not use inOrder for all verifications. For any verification that inOrder is not needed I just verified without inOrder. Unfortunately there are some unintended ordering are verified, like the ordering between the following
```java
    inOrder.verify(lb1).handleResolvedAddresses(resolvedAddressesCaptor1.capture());
    assertThat(resolvedAddressesCaptor1.getValue().getAddresses()).containsExactly(eag12);
    inOrder.verify(lb2).handleResolvedAddresses(resolvedAddressesCaptor2.capture());
    assertThat(resolvedAddressesCaptor2.getValue().getAddresses()).containsExactly(eag22);
    inOrder.verify(lb3).handleResolvedAddresses(resolvedAddressesCaptor3.capture());
    assertThat(resolvedAddressesCaptor3.getValue().getAddresses()).containsExactly(eag32);
    inOrder.verify(lb4).handleResolvedAddresses(resolvedAddressesCaptor4.capture());
    assertThat(resolvedAddressesCaptor4.getValue().getAddresses()).containsExactly(eag42);
```
but it's natural deterministic ordering (It's unlikely to change when implementation is changed). All unintended ordering verification is like this. I didn't simulate my implementation flow. I just followed the test case specification and verify all the expected behavior in the specification. If implementation changes and the unintended order verification breaks, we can change the test as long as it verifies all the test spec.

Reply:
What's your test specification for this test case? Precisely.